### ISSUE TITLE
RC2 Linking to invalid controller/action stops request

### ISSUE THREAD

**Issue by salerth** [aspnet/Mvc#3512]
Runtime: dnx-coreclr-darwin-x64.1.0.0-rc2-16128 (also impacts mono version)
Environment: OSX
MVC/Other deps : 6.0.0-rc2-16334 (6.0.0-*)

Previously if you added a `@Html.ActionLink("Home", "Index", "Home")` where the Controller/Action didnt exist, the link would be rendered empty without error.  With this update any links to invalid actions throws the following error:
```
fail: Microsoft.AspNet.Diagnostics.DeveloperExceptionPageMiddleware[0]
      An unhandled exception has occurred while executing the request
      System.NullReferenceException: Object reference not set to an instance of an object
        at System.Text.Encodings.Web.TextEncoder.Encode (System.IO.TextWriter output, System.String value) <0x10e5b2380 + 0x00018> in <filename unknown>:0 
        at Microsoft.AspNet.Mvc.Rendering.TagBuilder.AppendAttributes (System.IO.TextWriter writer, System.Text.Encodings.Web.HtmlEncoder encoder) <0x10e5d7980 + 0x001a1> in <filename unknown>:0 
        at Microsoft.AspNet.Mvc.Rendering.TagBuilder.WriteTo (System.IO.TextWriter writer, System.Text.Encodings.Web.HtmlEncoder encoder) <0x10e5d7580 + 0x00179> in <filename unknown>:0 
        at Microsoft.AspNet.Mvc.ViewFeatures.StringCollectionTextWriter+StringCollectionTextWriterContent.WriteTo (System.IO.TextWriter writer, System.Text.Encodings.Web.HtmlEncoder encoder) <0x10e5b6f10 + 0x001c0> in <filename unknown>:0 
        at Microsoft.AspNet.Mvc.ViewFeatures.StringCollectionTextWriter.CopyTo (System.IO.TextWriter writer, System.Text.Encodings.Web.HtmlEncoder encoder) <0x10e5c99f0 + 0x001b3> in <filename unknown>:0 
        at Microsoft.AspNet.Mvc.ViewFeatures.StringCollectionTextWriter.CopyToAsync (System.IO.TextWriter writer, System.Text.Encodings.Web.HtmlEncoder encoder) <0x10e5c99b0 + 0x0001e> in <filename unknown>:0 
        at (wrapper remoting-invoke-with-check) Microsoft.AspNet.Mvc.ViewFeatures.StringCollectionTextWriter:CopyToAsync (System.IO.TextWriter,System.Text.Encodings.Web.HtmlEncoder)
        at Microsoft.AspNet.Mvc.Razor.RazorTextWriter.CopyToAsync (System.IO.TextWriter writer) <0x10e5c9680 + 0x00068> in <filename unknown>:0 
        at Microsoft.AspNet.Mvc.Razor.RazorView+<RenderLayoutAsync>d__20.MoveNext () <0x10e512000 + 0x008d7> in <filename unknown>:0 
...
```

This error interupts the page rather than totally fails it, so the browser gets a `Net::ERR_INCOMPLETE_CHUNKED_ENCODING` (chrome) with a partially rendered page.

**Comment by rynowak** [comment]
The null-ref here is definitely unintended. 

As a side-discussion, we're planning to revert the change in behavior for link generation, and allow generating links to things that don't exist.

**Comment by GuardRex** [comment]
Moved from https://github.com/aspnet/Razor/issues/606

I just noticed that this markup ...
```html
<form id="contactform" asp-antiforgery="true" asp-action="contact" method="post">
```
... produces this markup ...
```html
<form id="contactform" method="post" action="">
```
... which, of course, hoses validation ...

![capture](https://cloud.githubusercontent.com/assets/1622880/11170138/b4571c9e-8b80-11e5-95a7-ba29cf8e96fb.PNG)

Is that the expected behavior?

@dougbu commented:

> @GuardRex this is somewhat expected because we do not currently generate links to non-existent actions. However we're discussing making changes in this area, as @rynowak mentioned in the aspnet/Mvc#3512 issue.

@dougbu Is there a simple work-around to get the page to validate that wouldn't break the processing of the form submission?

**Comment by dougbu** [comment]
@rynowak I made the suggestion to reuse this issue. If you think the two symptoms are unrelated, let's discuss breaking them apart tomorrow.

> Is there a simple work-around

@GuardRex for some reason `asp-action="contact"` isn't resolving. But you haven't provided enough information about the application to understand why. For example are you using attribute or template routing? Does a `Contact()` action exist in the executing controller? Does your code change any route values?

**Comment by GuardRex** [comment]
@dougbu Roasted rat burgers! :rat: :hamburger: You're right. I dorked it. I changed the Controller action several days ago, and thus the action element was empty when I forgot to update `asp-action="contact"`. My bad.

**Comment by dougbu** [comment]
@GuardRex np.

From my perspective, the silent failure is not helpful here. You would have known where to look sooner if either the `<form>` tag helper threw an `Exception` (as in @salerth's original post) or the generated `action` led to a 404 on submission.

**Comment by GuardRex** [comment]
@dougbu Thank you for some relief over it. I really, really don't want to waste anyone's time there with brain farts ... "code farts" let's call them.

The submission behavior was understandable ... a good route on the controller `[Route("/contact/{d?}")]` with a `[HttpPost]` seemed to mask it pretty good.

**Comment by rynowak** [comment]
BTW If you want to have MVC5's behavior here you can... https://github.com/aspnet/Mvc/blob/dev/test/WebSites/BestEffortLinkGenerationWebSite/Startup.cs

We're going to revert this change sometime in RC2, but until then here's a workaround.

**Comment by ctolkien** [comment]
From reading of the above, in an RC2 world, we can still re-enable this option by setting:

    options.UseBestEffortLinkGeneration = true;

? (we would prefer pages to error than fail to generate links).

**Comment by dougbu** [comment]
@ctolkien correct. That's what the `UseBestEffortLinkGeneration` option does.

**Comment by rynowak** [comment]
@ctolkien - we're simplifying the design here, and going to be consistent with MVC5 and previous versions.

If you want to throw when link-generation fails, then you can subclass UrlHelper and do your thing! We considered that, but didn't do it because it can be very disruptive to the way that people are used to authoring MVC.

Just be aware that we're reverting a design change to conventional routes. Conventional routes will soon behave similarly to when `UseBestEffortLinkGeneration` is always on, we're going to go back to MVC5's behavior.

**Comment by GuardRex** [comment]
@rynowak Me thinks this one is Announcement-worthy. If you post to Announcements, may I make a request to show the default behavior by example and the behavior with subclassing by example. I'm relatively new to MVC coming from years and years of Webforms, so examples and patterns are a big help trying to get my small brain wrapped around the concepts.

**Comment by rynowak** [comment]
@GuardRex - yes, there will be an example

> default behavior by example and the behavior with subclassing by example

I'm not sure what you mean. We're planning to make behavior of link generation compatible with MVC5 and previous versions.

The discussion about subclassing is related to how to handle cases where link generation does not succeed.

**Comment by GuardRex** [comment]
@rynowak I just meant to explain the statement ...

> If you want to throw when link-generation fails, then you can subclass UrlHelper and do your thing

... by example (i.e., code sample and behavior). It's obvious to folks who have been working for a while in MVC, and I realize to some extent its like "MVC 101" explaining something basic, but it would be helpful to see a short code sample and have an explanation of what "throw when link-generation fails" means (at compile time?) alongside what happens when link generation doesn't fail, i.e., `UseBestEffortLinkGeneration` is set to `false` (markup loses the value for `asp-action` if the controller doesn't exist?).

**Comment by rynowak** [comment]
> what happens when link generation doesn't fail, i.e., UseBestEffortLinkGeneration is set to false (markup loses the value for asp-action if the controller doesn't exist?).

#### Some basics

Link generation is a **runtime** feature of routing that lets you generate a URL using the routing system. This is the feature that backs `Url.Action(...)` `Html.Form(...)` and things like `RedirectToRouteResult` or `CreatedAtActionResult`. It's how we avoid hardcoding URLs in controllers and views.

Link generation runs through a list of routes in the route table, and tried each one in sequence to generate a URL. Each individual route has a chance when it is called to decide whether or not to generate a URL based on its own requirements.

If all routes decide not to generate a URL, then link generation 'fails' and routing returns null. This generally means that you get an empty URL in an `<a>` or something like that. (This has *always* been the MVC behavior when link generation fails).

#### The change in behavior

In MVC5, a route like `{controller}/{action}/{id?}` can generate a URL when you provide it with a value for `controller` and `action` (since id is marked as optional).

In MVC6, we tried an experimental feature to make this smarter. Now not only are `controller` and `action` required, but they have to match a real honest-to-goodness existing action method in your app in order for the URL to generate. Naturally this makes 'failed' link generation much more common.

The feedback we've gotten has been strongly in favor of the MVC5 behavior, so we're reverting it what we changed. The infrastructure that was required to build the MVC6 behavior was complex and requires some weird API, so we're pulling it out.

This is what the sample for throwing on failed link generation would look like:
```
public class MyUrlHelper : UrlHelper
{
    protected virtual string GeneratePathFromRoute(string routeName, IDictionary<string, object> values)
    {
        var result = base.GeneratePathFromRoute(routeName, values);
        if (result == null)
        {
             throw new MyException(<your message here>);
        }

        return result;
    }
}
```

`UseBestEffortLinkGeneration` will be removed as part of this wave of changes. This was a stopgap compromise to mostly give you MVC5's behavior.

**Comment by GuardRex** [comment]
@rynowak Perfect. Sorry for asking you to type out MVC 101 for Newbies! :baby: ... but that's a big help.

**Comment by rynowak** [comment]
This is fixed by 6875ee55d3c8632a92432b4b80b139f293f14804

We'll be removing `UseBestEffortLinkGeneration` in a future (soon) change.
